klasses/api.py

```diff
@@ -210,3 +212,121 @@ def fetch_bootcamp_run(run_property):
                     run_filters
                 )
             ) from exc
+
+
+def create_run_enrollments(user, runs, order=None):
+    """
+    Creates local records of a user's enrollment in bootcamp runs, and attempts to enroll them
+    in novoed via API
+
+    Args:
+        user (User): The user to enroll
+        runs (iterable of BootcampRun): The bootcamp runs to enroll in
+        order (ecommerce.models.Order or None): The order associated with these enrollments
+
+    Returns:
+        (list of BootcampRunEnrollment): A list of enrollment objects that were successfully
+            created
+    """
+    successful_enrollments = []
+
+    for run in runs:
+        try:
+            enrollment, created = BootcampRunEnrollment.objects.get_or_create(
+                user=user, bootcamp_run=run
+            )
+            if not created and not enrollment.active:
+                enrollment.reactivate_and_save()
+            elif order:  # enrollment created and order is available
+                application = order.application
+                application.bootcamp_run = run
+                application.save()
+
+            if (
+                features.is_enabled(features.NOVOED_INTEGRATION)
+                and enrollment.bootcamp_run.novoed_course_stub
+            ):
+                novoed_tasks.enroll_users_in_novoed_course.delay(
+                    user_id=enrollment.user.id,
```

Comment:
This code is functionally very similar to `applications.models.BootcampApplication.complete`, which makes sense. Having 2 copies of this logic opens us up to bugs in a very critical area of the application. We should move this logic to a separate method so there's just one place where we handle a user completing enrollment in a bootcamp run. Also, we should create a separate method for enrolling a user in a single run (`create_run_enrollments` just performs the enrollment actions and aggregates the results).

Changes would look something like this:

```python
# models.py
class BootcampApplication(TimestampedModel):
  def complete(self):
    create_run_enrollment(self.user, self.bootcamp_run)

# klasses/api.py
def defer_enrollment(
    user, from_bootcamp_run_id, to_bootcamp_run_id, order_id, force=False
):
  # ...
  to_enrollments = create_run_enrollment(user, to_run, order=order)

def create_run_enrollment(user, run, order=None):
    enrollment, _ = BootcampRunEnrollment.objects.update_or_create(
        user=user,
        bootcamp_run=run,
        defaults={"active": True, "change_status": None},
    )
    if order:
        application = order.application
        application.bootcamp_run = run
        application.save()
    if (
        features.is_enabled(features.NOVOED_INTEGRATION)
        and run.novoed_course_stub
    ):
        novoed_tasks.enroll_users_in_novoed_course.delay(
            user_id=user.id,
            novoed_course_stub=run.novoed_course_stub,
        )
    return enrollment

def create_run_enrollment(user, run, order=None):
    successful_enrollments = []

    for run in runs:
        try:
            enrollment = create_run_enrollment(user, run, order=None)
    # ...
    return successful_enrollments
```